### PR TITLE
make consent service dont call willSeeResource on children if parent resource is AUTHORIZED or REJECT

### PR DESCRIPTION
**Pull request by mrdnctrk**
This PR makes sure that when retrieving a resource that may contain other resources (a bundle for example), if a ConsentService returns AUTHORIZED or REJECT for that parent resource on willSeeResource call, the willSeeResource is not called for the child resources contained in the parent resource. 

The ConsentInterceptor class (which calls the ConsentServices) had already in place this logic of not calling willSeeResource on the child Resources upon getting AUTHORIZED or REJECT from the parent, but it was working for the bundles returned for the Search results. It wasn't working when parent resource is retrieved directly by its id, for example a Bundle of type document that is persisted in the DB, not a search result bundle that is generated by the server. 

The main reason is why this wasn't working for persisted Bundles is as follows. ConsentInterceptor calls willSeeResource on two pointcuts in this flow: 
- first in STORAGE_PRESHOW_RESOURCES hook
- second in SERVER_OUTGOING_RESPONSE hook 

When retrieving a persisted bundle, the bundle is processed by the STORAGE_PRESHOW_RESOURCES hook, which sends the bundle to ConsentServices by calling willSeeResource. Then SERVER_OUTGOING_RESPONSE is invoked, which in the existing code was checking that the bundle resource was already processed and it was directly starting calling willSeeResource on the child resources. 

It seems that the reason for this existing logic is that on server generated bundles returned in the response ( for example for a bundle returning the search results) STORAGE_PRESHOW_RESOURCES is not called for the bundle,  but called for the child resources that is being returned in the bundle. This makes sense because the bundle is server generated when returning response (not an actual resource in the db). So the bundle is not available on STORAGE_PRESHOW_RESOURCES yet. The search result bundle itself is processed in 
SERVER_OUTGOING_RESPONSE for the first time, and the existing code detects this and does not call willSeeResource on the child resources. 

In summary, the existing code assumed that if we are returning a bundle, it will be processed for the first time in SERVER_OUTGOING_RESPONSE but not in STORAGE_PRESHOW_RESOURCES, and this assumption is wrong for persisted bundles. For a persisted bundle, STORAGE_PRESHOW_RESOURCES processes the bundle the first.
 
My change is  when STORAGE_PRESHOW_RESOURCES processes a resource, I store the ConsentService decision for the resource. When SERVER_OUTGOING_RESPONSE is called, I check if the resource was processed already and if it was and the decision was AUTHORIZED or REJECT, I make the function return immediately without calling willSeeResource on the children. (Otherwise if the decision was PROCEED or the resource wasn't processed before, I let the function run as before).

### CODE REVIEW

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java [hunk 1]
```diff
@@ -508,8 +507,7 @@ private void checkOutgoingResourceAndFailIfDeny(
 		}
 
 		// Don't check the value twice
-		IdentityHashMap<IBaseResource, Boolean> alreadySeenMap =
-				ConsentInterceptor.getAlreadySeenResourcesMap(theRequestDetails, myRequestSeenResourcesKey);
```
Comment (mrdnctrk) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java:512]: this Interceptor had used the static method on ConsentIntercetor to get its own already seen map ( the key used to retrieve the map from the userData is specific to this interceptor so it is not related to ConsentInterceptor). I moved the logic of retrieving the map to this class instead of using that static method from ConsentInterceptor, which I have changed.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/ConsentInterceptor.java [hunk 2]
```diff
@@ -441,12 +468,17 @@ public boolean acceptElement(
...
+								replacementResource = childOutcome.getResource();
+								shouldReplaceResource = replacementResource != null;
+								break;
 							case AUTHORIZED:
 								replacementResource = childOutcome.getResource();
 								shouldReplaceResource = replacementResource != null;
-								shouldCheckChildren &= childOutcome.getStatus() == ConsentOperationStatusEnum.PROCEED;
+								shouldCheckChildren = false;
+								alreadySeenResources.put(resource, ConsentOperationStatusEnum.AUTHORIZED);
 								break;
```
Comment (mrdnctrk) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/consent/ConsentInterceptor.java:482]: Previously in this switch PROCEED and AUTHORIZED cases were handled together, I separated them into two. 

In the new version, the common part of the logic
```
replacementResource = childOutcome.getResource();
shouldReplaceResource = replacementResource != null;
``
is repeated for both cases as is.  

The old version also had: 
`shouldCheckChildren &= childOutcome.getStatus() == ConsentOperationStatusEnum.PROCEED;`
and this is now separated as well. 

For `PROCEED` case, this part basically means to keep the existing value `shouldCheckChildren` ( because `shouldCheckChildren = shouldCheckChildren & true`) will not change the value of the `shouldCheckChildren` (if it was false it will remain false, if it was true it will remain true)

For `AUTHORIZED` case, this part means make `shouldCheckChildren` false. (because `shouldCheckChildren = shouldCheckChildren & false`) is always assigns false shouldCheckChildren.

**Review by TipzCM — APPROVED** [review]

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java [hunk 3]
```diff
@@ -678,4 +676,15 @@ private String getResourceTypeOrEmpty(IBaseResource theResource) {
...
 		return theResource.getIdElement().getResourceType();
 	}
+
+	@SuppressWarnings("unchecked")
+	private IdentityHashMap<IBaseResource, Boolean> getAlreadySeenResourcesMap(RequestDetails theRequestDetails) {
+		IdentityHashMap<IBaseResource, Boolean> alreadySeenResources = (IdentityHashMap<IBaseResource, Boolean>)
+				theRequestDetails.getUserData().get(myRequestSeenResourcesKey);
+		if (alreadySeenResources == null) {
+			alreadySeenResources = new IdentityHashMap<>();
+			theRequestDetails.getUserData().put(myRequestSeenResourcesKey, alreadySeenResources);
```
[the remark this answers is not in the document]
Comment (mrdnctrk) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/interceptor/auth/AuthorizationInterceptor.java:686]: yes, it can be made static.

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/HashMapResourceProvider.java [hunk 4]
```diff
@@ -580,7 +580,10 @@ private static <T extends IBaseResource> T fireInterceptorsAndFilterAsNeeded(
 		List<IBaseResource> output =
 				fireInterceptorsAndFilterAsNeeded(Lists.newArrayList(theResource), theRequestDetails);
 		if (output.size() == 1) {
-			return theResource;
+			// do not return theResource here but return whatever the interceptor returned in the list because
+			// the interceptor might have set the resource in the list to null (if it didn't want it to be returned).
+			// ConsentInterceptor might do this for example.
+			return (T) output.get(0);
```
[the remark this answers is not in the document]
Comment (mrdnctrk) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/provider/HashMapResourceProvider.java:586]: I am not really sure, but it looks like interceptors cannot remove elements from the list, but they can set the list elements to null as ConsentInterceptor does. If the list doesn't have any elements there is the else part already which returns null.
